receiver.go

```diff
@@ -408,33 +412,43 @@ func (r *Receiver) newSessionAndLink(ctx context.Context) error {
 		opts = append(opts, amqp.LinkSenderSettle(amqp.ModeSettled))
 	}
 
-	if opt, ok := r.getSessionFilterLinkOption(); ok {
-		opts = append(opts, opt)
+	sessionOpt, useSessionOpt := r.getSessionFilterLinkOption()
+	if useSessionOpt {
+		opts = append(opts, sessionOpt)
 	}
 
 	amqpReceiver, err := amqpSession.NewReceiver(opts...)
 	if err != nil {
 		tab.For(ctx).Error(err)
 		return err
 	}
-
 	r.receiver = amqpReceiver
+	if useSessionOpt {
+		rawsid := r.receiver.LinkSourceFilterValue(sessionFilterName)
+		if rawsid == nil && r.sessionID == nil {
+			return errors.New("failed to create a receiver.  no unlocked sessions available")
+		} else if rawsid != nil && r.sessionID != nil && rawsid != *r.sessionID {
+			return fmt.Errorf("failed to create a receiver for session %s, it may be locked by another receiver", rawsid)
+		} else if r.sessionID == nil {
+			sid := rawsid.(string)
+			r.sessionID = &sid
+		}
+	}
 	return nil
 }
 
 func (r *Receiver) getSessionFilterLinkOption() (amqp.LinkOption, bool) {
-	const name = "com.microsoft:session-filter"
 	const code = uint64(0x00000137000000C)
 
 	if !r.useSessions {
 		return nil, false
 	}
 
 	if r.sessionID == nil {
-		return amqp.LinkSourceFilter(name, code, nil), true
+		return amqp.LinkSourceFilter(sessionFilterName, code, nil), true
```

Comment:
Looks like you don't need the 'if' here at all (ie, r.SessionID can just be passed unconditionally)

Reply:
This is one of the cases of "when nil isn't nil" in Go, and comes into play with `interface{}`.  When passing `nil`, the type of the `value interface{}` param will be `nil`.  When passing `r.sessionID`, the type will be `*string`.  So while they both could in fact be `nil`, their underlying types are different.  This is important to the [marshal](https://github.com/Azure/go-amqp/blob/master/encode.go#L43) function in `go-amqp` which differentiates between underlying types.